### PR TITLE
demo rake script

### PR DESCRIPTION
**Pull request by jpslav**
This should be good to go, at least for FE to start playing with. 

`bundle exec rake db:drop db:create db:migrate db:seed demo:001`

Outputs something like:

```
Imported book e4c329f3-1972-4835-a203-3e8c539e4df3@2.14 from https://archive-staging-tutor.cnx.org/contents/ and https://exercises-demo.openstax.org.
Added Bill Nye as a teacher and added 20 students.
Assignments will be due around 2015-05-13 00:01:57 -0700.
Assigned ireading for ["3.1", "3.2", "3.3"], due: 2015-05-08 00:01:57 -0700, title: Introduction
One task looks like: Task 1 / reading / 1r 2r 3e 4r 5r 6r 7e 8v 9e 10r 11r 12i 13e 14r 15r 16e 17r 18r 19r 20e 21r 22e 23r
Assigned homework, 'Homework - 3.1-3.2-3.3' due at 2015-05-10 07:01:57 UTC; 20 times
One task looks like: Task 21 / homework / 461e 462e 463e 464e 465e 466e 467e 468e 469e 470e
Assigned ireading for ["4.1", "4.2"], due: 2015-05-13 00:01:57 -0700, title: Introduction
One task looks like: Task 41 / reading / 661r 662r 663e
Assigned homework, 'Homework - 4.1-4.2' due at 2015-05-13 07:01:57 UTC; 20 times
One task looks like: Task 61 / homework / 721e 722e 723e 724e 725e 726e 727e 728e 729e
Success!
```

The "One task looks like" lines list task steps at the end in the format "#{step id}#{step type first letter}".

This script creates 2 iReadings and 2 HWs, and attempts to work them to achieve the grades set out by UX (which is not always achievable because we don't always have that many exercises).

The users created by this script have the terms of use and privacy policy already agreed to.

I haven't tested this much beyond manual poking around.  Looking forward to hearing what FE has to say.

Right now the dates are centered around the current day.  Later when we get time traveling folded in, we'll move the dates out to October.

/cc @philschatz @pandafulmanda @nathanstitt @jivewise @darrenmason @dak

### CODE REVIEW

File: app/external/openstax/cnx/v1.rb [hunk 1]
```diff
@@ -2,12 +2,34 @@
 
 module OpenStax::Cnx::V1
 
-  ARCHIVE_URL_BASE ='http://archive.cnx.org/contents/'
-  SSL_ARCHIVE_URL_BASE ='https://archive.cnx.org/contents/'
+  # Sets the archive URL base.  'url' is nominally the non-SSL URL,
+  # tho may be SSL.  An explicit SSL URL can be passed in `ssl`, or
+  # by default the SSL URL will be guessed from the URL.
+  def self.set_archive_url_base(url:, ssl: :automatic)
+    @archive_url_base = url
```
Comment (Dantemss) [inline app/external/openstax/cnx/v1.rb]: Might want to automatically convert https to http for this one. Also, why do we need both?

Comment (jpslav) [inline app/external/openstax/cnx/v1.rb]: Maybe not important now, unless we run into some case where a dev machine doesn't have SSL?

**Comment by jpslav** [comment]
@Dantemss please review / merge.

**Comment by jpslav** [comment]
Right now I don't track which are the core steps, so easiest way to make sure the core was incomplete was to make the first one incomplete (others may be incomplete too)

**Comment by jpslav** [comment]
But I guess you're right -- `:incomplete` doesn't need to mean the core is incomplete -- could make the last one incomplete -- I'll do that.

**Comment by philschatz** [comment]
Using student01 for these screenshots/issues:

Broken image URL (maybe bad content?): 
image

Should there be upcoming tasks for the students?
image

Should exercises in the middle of an iReading have a `<<` icon and label?
image

**Comment by jpslav** [comment]
Re "upcoming tasks", I don't think there are, at least per my interpretation of the google doc description (with an initial date of Oct 14) combined with Jason's description of which students have worked what:

image

I think the upcoming tasks will be populate when the teacher makes new assignments during the demo.

**Comment by pandafulmanda** [comment]
in reference to the last screenshot, I'm confused.  That is based on it being a spaced practice from the BE.  I think we are getting a new icon to represent them.  Should these not be showing?  or should spaced practices not be in the middle of a reading?

**Comment by jpslav** [comment]
Re the "broken image URL" -- that appears to be in the actual content: 

https://exercises-demo.openstax.org/api/exercises/?q=id:33@1

I don't think we have the ability to upload images to exercises yet via the spreadsheet -- @Dantemss can correct me if I'm wrong.  

@pandafulmanda I think Phil is not talking about the pencil icon but some double arrow icon... but I don't know what he means by that either :-)

**Comment by philschatz** [comment]
@pandafulmanda The `<<` icon shows up for the first exercise in a reading (I assume because it has a field in the JSON for `chapter_section`) but I think UX only intended those to be for spaced-practice exercises. So, either BE would have to signal that an exercise step is spaced (my pref) or not include that `chapter_section` info for the non-spaced exercise.

Unless of course BE has already done that, and I apologize for not remembering the updates to the API :innocent:

**Comment by jpslav** [comment]
@philschatz that exercise is actually the spaced practice exercise for that reading -- that reading doesn't have any other exercises

```ruby
{
  id: "663",
  task_id: "41",
  type: "exercise",
  group: "spaced practice",
  is_completed: false,
  related_content: [
    {
      title: "Acceleration",
      chapter_section: "3.2"
    }
  ],
```

**Comment by philschatz** [comment]
@jpslav I have **no idea** what you guys are talking about

**Comment by jpslav** [comment]
Funny funny

**Comment by pandafulmanda** [comment]
:laughing:

**Comment by jpslav** [comment]
I think @kevinburleigh75 intentionally didn't use an underscore; if there's a reason to have it we can.

File: spec/external/openstax/cnx/v1/page_spec.rb [hunk 2]
```diff
@@ -201,4 +201,29 @@
...
+      OpenStax::Cnx::V1::Fragment::Text,
+      OpenStax::Cnx::V1::Fragment::Interactive,
+      OpenStax::Cnx::V1::Fragment::Exercise,
+      OpenStax::Cnx::V1::Fragment::Text,
+      OpenStax::Cnx::V1::Fragment::Text,
+      OpenStax::Cnx::V1::Fragment::Text,
+      OpenStax::Cnx::V1::Fragment::ExerciseChoice,
+      OpenStax::Cnx::V1::Fragment::Video,
+      OpenStax::Cnx::V1::Fragment::Exercise,
+      # OpenStax::Cnx::V1::Fragment::Exercise,  # put back after merge fixes
```
Comment (Dantemss) [inline spec/external/openstax/cnx/v1/page_spec.rb]: ? :)

**Comment by kevinburleigh75** [comment]
Whether or not the `group` fields have underscores or spaces is easy enough to change if we have a strong opinion one way or the other.

**Comment by Dantemss** [comment]
Might want a spec like the one for setup 001 just to make sure the demo task runs.

**Comment by Dantemss** [comment]
https://github.com/openstax/tutor-server/blob/master/spec/lib/tasks/setup_001_spec.rb

File: spec/external/openstax/cnx/v1/page_spec.rb (hunk 2 above)
Comment (jpslav) [inline spec/external/openstax/cnx/v1/page_spec.rb]: That's after the code goes in that deals with multiple exercises in a grasp check... the currently blocked PR https://github.com/openstax/tutor-server/pull/259.  There are two exercises in the content but we're currently only picking up one.

**Comment by philschatz** [comment]
@kevinburleigh75 it just seems a bit odd to have spaces in non-human-readable strings and raises questions like:

- are those 2 values or 1?
- when to use spaces and when to use underscores?

File: spec/lib/tasks/demo_001_spec.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+require 'vcr_helper'
+require 'tasks/demo_001'
+
+RSpec.describe Demo001, type: :request, version: :v1, speed: :slow, vcr: VCR_OPTS do
+
+  it "doesn't catch on fire" do
+    expect(Demo001.call.errors).to be_empty
+  end
+
+end
```
Comment (jpslav) [inline spec/lib/tasks/demo_001_spec.rb]: @Dantemss @Dantemss @Dantemss starts with @D and he might need new glasses.

**Comment by Fredasaurus** [comment]
@philschatz I think having it it helpful. It affirms that you have nothing to do, rather than having to assume it.

**Comment by kevinburleigh75** [comment]
@philschatz Ok - I will make a note to change the spaces in `group` to underscores.

File: spec/lib/tasks/demo_001_spec.rb (hunk 3 above)
Comment (Dantemss) [inline spec/lib/tasks/demo_001_spec.rb]: (doh)

File: spec/external/openstax/cnx/v1/page_spec.rb (hunk 2 above)
Comment (Dantemss) [inline spec/external/openstax/cnx/v1/page_spec.rb]: ah ok

**Pull request merged**